File: day11/solve.ts
import { solve } from "../runner/typescript";
import { sum, range } from "lodash";

type Star = readonly [number, number];

function parser(input: string): Star[] {
  return input
    .split("\n")
    .flatMap((line, y) =>
      line.split("").map((char, x) => (char == "#" ? ([x, y] as const) : null))
    )
    .filter(Boolean);
}

function findDistance(stars: Star[], expansion: number) {
  const pairs = stars.flatMap((a, i) => stars.slice(i + 1).map((b) => [a, b]));

  const distances = pairs.map(([[ax, ay], [bx, by]]) => {
    const [minY, maxY] = [Math.min(ay, by), Math.max(ay, by)];
    const [minX, maxX] = [Math.min(ax, bx), Math.max(ax, bx)];
    const starlessRowsBetween = range(minY, maxY).filter(
      (y) => !stars.some((s) => s[1] == y)
    );
    const starlessColsBetween = range(minX, maxX).filter(
      (x) => !stars.some((s) => s[0] == x)
    );
    return (
      Math.abs(by - ay) +
      Math.abs(bx - ax) +
      starlessRowsBetween.length * (expansion - 1) +
      starlessColsBetween.length * (expansion - 1)
    );
  });

  return sum(distances);
}

function part1(stars: Star[]) {
  return findDistance(stars, 2);
}

function part2(stars: Star[], isTest: boolean): number {
  return findDistance(stars, isTest ? 100 : 1000000);
}

const testInput =
  "...#......\n.......#..\n#.........\n..........\n......#...\n.#........\n.........#\n..........\n.......#..\n#...#.....";

solve({
  parser,
  part1,
  part2,
  part1Tests: [[testInput, 374]],
  part2Tests: [[testInput, 8410]],
});

File: day11/solve.py
from runner.python import solve
from itertools import combinations


def parse(input: str) -> list[tuple[int, int]]:
  return [
    (x, y)
    for y, line in enumerate(input.split("\n"))
    for x, char in enumerate(line)
    if char == "#"
  ]


def find_distance(stars: list[tuple[int, int]], expansion: int) -> int:
  xs, ys = {x for x, _ in stars}, {y for _, y in stars}

  return sum(
    abs(by - ay)
    + abs(bx - ax)
    + (expansion - 1) * len(set(range(ay, by)) - ys)
    + (expansion - 1) * len(set(range(*sorted([ax, bx]))) - xs)
    for (ax, ay), (bx, by) in combinations(stars, 2)
  )


def part1(input):
  return find_distance(parse(input), 2)


def part2(input):
  return find_distance(parse(input), 1000000)


testInput = "...#......\n.......#..\n#.........\n..........\n......#...\n.#........\n.........#\n..........\n.......#..\n#...#....."
solve(
  part1,
  testInput,
  374,
  part2,
  testInput,
  82000210,
)
